other/upgrade.php

```diff
@@ -1401,13 +1427,22 @@ function UpgradeOptions()
 	// Ensure this doesn't get lost in translation.
 	$changes['upgradeData'] = base64_encode(json_encode($upcontext['user']));
 
+	// Update $auth_secret?
+	if (!empty($_POST['resetAuthSecret']))
+		$changes['auth_secret'] = bin2hex($smcFunc['random_bytes'](32));
+
```

Comment:
I thought it would be good option as upgrades between major versions don't happen often, and rotating out secrets like this can be beneficial to security.

Reply:
The problem is, changing the $auth_secret value doesn't just invalidate existing cookies; it also invalidates all of the unsubscribe links we've included our email notifications ever since 2.0.16 was released and since #6012 was merged for 2.1. Those new unsubscribe links are part of GDPR compliance, since the GDPR requires us to allow someone to unsubscribe their email address from a given type of notification *without needing to log into an account.* (Perhaps that's a silly requirement, but it is what it is.) But the only way to meet that requirement even halfway securely is to include a unique token in the URL parameters, and the $auth_secret value is integral to generating and verifying those tokens. So if $auth_secret changes, the unsubscribe links break, and thus the forum runs afoul of the GDPR.

The need to keep $auth_secret invariant once created is why there's no UI to change it in the Admin Centre and why the comments in Settings.php itself strongly warn the admin against changing it manually. For the same reason, we shouldn't offer to change it during upgrade.

Reply:
Should we do the job right then in 2.1 and split this into a secondary secret used for just notification links?
We could add a note that this will break logins and invalidate all notification unsubscribe links which may be in violation of a GDPR or similar requirements.

If we do it for notifications, we could store it in the database.  So we could store a notifications_secret and a notifications_secret_old, allowing the old secret to work for a period of time.

Preventing rotating of secrets is a bad thing.  While it allows it to be unique among all forums, it prevents security related actions of rotating those secrets on intervals, which may be another checkbox on some security compliance form somewhere.

Reply:
Splitting it into two secrets wouldn't get us any further. It would merely leave us with two secrets that would both still be subject to the same requirements that one secret would. They'd be used in almost identical ways and stored in identical ways (neither one could be stored in the database, or else a database leak would give the attacker all the info they need to forge the login authentication, so they'd have to be next to each other in Settings.php), so any policy requiring one of them to be periodically changed would equally require the other to be periodically changed.

Storing old versions of the secret and checking against is possible in theory, I suppose, but it wouldn't accomplish much. If old keys still exist and can be used to validate incoming links or cookies, then they're still active and haven't actually been rotated out.

But regardless of the above, I don't believe there's any need to rotate this particular key anyway. It never, ever leaves the server in any form. It can't be extracted from any of the tokens or cookie values we generate and send out. The only way for an attacker to get ahold of it is to gain access to the server file system itself so they can read Settings.php, at which point it's already far too late.